src/compiler/crystal/codegen/primitives.cr

```diff
@@ -1011,9 +1010,6 @@ class Crystal::CodeGenVisitor
       target_def.abi_info = old_abi_info
       target_def.c_calling_convention = old_c_calling_convention
     end
-
-    old_needs_value = @needs_value
```

Comment:
This seems more like a hidden bug to me. This line should've been `@needs_value = old_needs_value` probably.
@asterite  WDYT?

Reply:
Indeed!

Reply:
I've pushed commit with this fix.

Reply:
Thanks, but this fix should be in a separate PR. Ideally, with a spec.